File: src/main/java/no/nav/samordning/hendelser/security/TokenResolver.java
package no.nav.samordning.hendelser.security;

import com.auth0.jwt.JWT;
import no.nav.samordning.hendelser.consumer.TpregisteretConsumer;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.server.resource.BearerTokenError;
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes;
import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver;

import javax.servlet.http.HttpServletRequest;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Collectors;

public class TokenResolver implements BearerTokenResolver {

    private static final Logger LOG = LoggerFactory.getLogger(TokenResolver.class);
    private static final String REQUIRED_SCOPE = "nav:pensjon/v1/samordning";

    private static final List<String> REQUIRED_CLAIM_KEYS = List.of(
            ClaimKeys.CLIENT_ID,
            ClaimKeys.CLIENT_ORGANISATION_NUMBER,
            ClaimKeys.ISSUER,
            ClaimKeys.SCOPE);

    private final BearerTokenResolver bearerTokenResolver;
    private final TpregisteretConsumer tpRegisteretConsumer;
    private final String serviceUser;
    private final String serviceUserIssuer;

    public TokenResolver(
            BearerTokenResolver bearerTokenResolver,
            TpregisteretConsumer tpRegisteretConsumer,
            String serviceUser,
            String serviceUserIssuer) {
        this.bearerTokenResolver = bearerTokenResolver;
        this.tpRegisteretConsumer = tpRegisteretConsumer;
        this.serviceUser = serviceUser;
        this.serviceUserIssuer = serviceUserIssuer;
    }

    @Override
    public String resolve(HttpServletRequest request) {
        var token = bearerTokenResolver.resolve(request);

        if (token == null) {
            return null;
        }

        var claims = getClaims(token);
        var tpnr = request.getParameter("tpnr").split("\\?")[0];

        if (validServiceUser(claims)) {
            LOG.info("Valid service user token");
            claims.put(ClaimKeys.CLIENT_ORGANISATION_NUMBER, "00000000");
            if (validOrganisation(tpnr, claims)) {
                LOG.info("Valid orgnr mapping");
            } else {
                LOG.info("Invalid orgnr mapping");
            }
            return token;
        }

        checkThatRequiredParametersAreProvided(claims);

        if (validScope(claims) && validOrganisation(tpnr, claims)) {
            log("Valid", tpnr, claims);
            return token;
        }

        log("Invalid", tpnr, claims);
        LOG.info("Invalid token");
        return null;
    }

    private void checkThatRequiredParametersAreProvided(Map<String, Object> claims) {
        String missingKeys = REQUIRED_CLAIM_KEYS
                .stream()
                .filter(key -> !claims.containsKey(key))
                .collect(Collectors.joining(", "));

        if (missingKeys.length() > 0) {
            throw tokenError("Missing parameters: " + missingKeys);
        } else {
            LOG.info("Client_ID: " + claims.get(ClaimKeys.CLIENT_ID).toString());
            LOG.info("ORNGNR: " + claims.get(ClaimKeys.CLIENT_ORGANISATION_NUMBER).toString());
            LOG.info("SCOPE: " + claims.get(ClaimKeys.SCOPE).toString());
            LOG.info("ISSUER: " + claims.get(ClaimKeys.ISSUER).toString());
        }
    }

    private boolean validOrganisation(String tpnr, Map<String, Object> claims) {
        String organisationNumber = claims.get(ClaimKeys.CLIENT_ORGANISATION_NUMBER).toString();
        return tpRegisteretConsumer.validateOrganisation(organisationNumber, tpnr);
    }

    private boolean validScope(Map<String, Object> claims) {
        var validScope = REQUIRED_SCOPE.equals(claims.get(ClaimKeys.SCOPE).toString());
        if (!validScope) {
            LOG.info("Invalid scope: " + claims.get(ClaimKeys.SCOPE).toString());
        }
        return validScope;
    }

    private boolean validServiceUser(Map<String, Object> claims) {
        return claims.containsKey(ClaimKeys.AUTHORIZED_PARTY)
                && claims.containsKey(ClaimKeys.ISSUER)
                && claims.get(ClaimKeys.AUTHORIZED_PARTY).toString().equals(serviceUser)
                && claims.get(ClaimKeys.ISSUER).toString().equals(serviceUserIssuer);
    }

    private static Map<String, Object> getClaims(String token) {
        var base64Payload = JWT.decode(token).getPayload();
        var payload = new String(Base64.getUrlDecoder().decode(base64Payload), StandardCharsets.UTF_8);
        return asMap(new JSONObject(payload));
    }

    private static Map<String, Object> asMap(JSONObject json) {
        var map = new HashMap<String, Object>();
        var keys = json.keys();

        while (keys.hasNext()) {
            var key = keys.next();
            map.put(key, json.get(key));
        }

        return map;
    }

    private static OAuth2AuthenticationException tokenError(String message) {
        var error = new BearerTokenError(BearerTokenErrorCodes.INVALID_REQUEST, HttpStatus.BAD_REQUEST, message,
                "https://tools.ietf.org/html/rfc6750#section-3.1");
        return new OAuth2AuthenticationException(error);
    }

    private static void log(String status, String tpnr, Map<String, Object> claims) {
        LOG.info(String.format("%s tpnr %s for client_id %s", status, tpnr, claims.get(ClaimKeys.CLIENT_ID)));
    }

    private class ClaimKeys {
        static final String AUTHORIZED_PARTY = "azp";
        static final String CLIENT_ID = "client_id";
        static final String CLIENT_ORGANISATION_NUMBER = "client_orgno";
        static final String ISSUER = "iss";
        static final String SCOPE = "scope";
    }
}

File: src/main/kotlin/no/nav/samordning/hendelser/security/TokenResolver.kt
package no.nav.samordning.hendelser.security

import com.auth0.jwt.JWT
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import no.nav.samordning.hendelser.consumer.TpConfigConsumer
import org.slf4j.Logger
import org.slf4j.LoggerFactory.getLogger
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpStatus.BAD_REQUEST
import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.oauth2.server.resource.BearerTokenError
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes.INVALID_REQUEST
import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver
import org.springframework.security.oauth2.server.resource.web.DefaultBearerTokenResolver
import org.springframework.stereotype.Service
import java.util.*

@Service
class TokenResolver(
    private val tpConfigConsumer: TpConfigConsumer,
    @Value("\${oauth2.acceptAll:#{null}}") private val acceptAll: String?,
    private val mapper: ObjectMapper
) : BearerTokenResolver {
    private val log = getLogger(javaClass)

    internal var bearerTokenResolver: BearerTokenResolver = DefaultBearerTokenResolver()

    init {
        log.info("Granting access to all resources for orgno ${acceptAll ?: "N/A"}.")
    }

    override fun resolve(request: HttpServletRequest) = bearerTokenResolver.resolve(request)?.takeIf {
        validateClaims(
            getClaims(it), request.getParameter("tpnr").substringBefore('?')
        )
    }

    private fun validateClaims(claims: JsonNode, tpnr: String) =
        claims.validScope().and(claims.validOrganisation(tpnr)).also {
            log(if (it) "Valid" else "Invalid", tpnr, claims)
            if (!it) log.info("Invalid token")
        }

    private fun JsonNode.validOrganisation(tpnr: String) = orgNo == acceptAll || tpConfigConsumer.validateOrganisation(
        orgNo, tpnr
    )!!

    private fun getClaims(token: String) =
        mapper.readTree(decode(token)).also(::checkThatRequiredParametersAreProvided)

    private fun decode(token: String) = String(
        Base64.getUrlDecoder().decode(
            JWT.decode(token).payload
        )
    )

    private fun checkThatRequiredParametersAreProvided(claims: JsonNode) {
        REQUIRED_CLAIM_KEYS.filterNot(claims::has).joinToString(", ").takeUnless(String::isEmpty)
            ?.let { throw tokenError("Missing parameters: $it") } ?: log.logClaims(claims)
    }

    private fun Logger.logClaims(claims: JsonNode) {
        info("Client_ID: ${claims[CLIENT_ID].asText()}")
        info("ORNGNR: ${claims.orgNo}")
        info("SCOPE: ${claims[SCOPE].asText()}")
        info("ISSUER: ${claims[ISSUER].asText()}")
    }

    private fun JsonNode.validScope() = this[SCOPE].asText().split(' ').let {
        if (REQUIRED_SCOPE in it) true
        else {
            log.info("Invalid scope: $it")
            false
        }
    }

    private val JsonNode.orgNo: String
        get() = stsOrgNo ?: supplier ?: consumer ?: ""

    private val JsonNode.stsOrgNo: String?
        get() = get(CLIENT_ORGANISATION_NUMBER)?.takeUnless { it.isMissingNode }?.asText()

    private val JsonNode.supplier: String?
        get() = get(SUPPLIER)?.takeUnless { it.isMissingNode }?.get(ID)?.takeUnless { it.isMissingNode }?.asText()
            ?.substringAfterLast(':')

    private val JsonNode.consumer: String?
        get() = get(CONSUMER)?.takeUnless { it.isMissingNode }?.get(ID)?.takeUnless { it.isMissingNode }?.asText()
            ?.substringAfterLast(':')

    private fun tokenError(message: String) = OAuth2AuthenticationException(
        BearerTokenError(
            INVALID_REQUEST, BAD_REQUEST, message, "https://tools.ietf.org/html/rfc6750#section-3.1"
        )
    )

    private fun log(status: String, tpnr: String, claims: JsonNode) =
        log.info("$status tpnr $tpnr for client_id ${claims[CLIENT_ID].asText()}")

    companion object {
        private const val REQUIRED_SCOPE = "nav:pensjon/v1/samordning"
        private const val CLIENT_ID = "client_id"
        private const val CLIENT_ORGANISATION_NUMBER = "client_orgno"
        private const val CONSUMER = "consumer"
        private const val SUPPLIER = "supplier"
        private const val ID = "ID"
        private const val ISSUER = "iss"
        private const val SCOPE = "scope"

        private val REQUIRED_CLAIM_KEYS = listOf(
            CLIENT_ID, ISSUER, SCOPE
        )
    }
}
